subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                            false
+                        }
+                    }
+                };
+
+                if verified {
+                    info!("proof for {} verified!", txid);
+                    self.db.insert_tx_verified(&txid)?;
+                    found += 1;
+                }
+            }
+        }
+        Ok(found)
+    }
+
+    pub fn remove(&mut self, headers: u32) -> Result<(), Error> {
+        if let ChainOrVerifier::Chain(chain) = &mut self.checker {
+            chain.remove(headers)?;
+        }
+        Ok(())
+    }
+}
+impl<S: Read + Write> Syncer<S> {
+    pub fn sync(&mut self) -> Result<bool, Error> {
+        trace!("start sync");
+        let start = Instant::now();
+
+        let mut history_txs_id = HashSet::new();
+        let mut heights_set = HashSet::new();
+        let mut txid_height = HashMap::new();
+
+        let mut last_used = [0u32; 2];
+        for i in 0..=1 {
+            let int_or_ext = Index::from(i)?;
+            let mut batch_count = 0;
+            loop {
+                let batch = self.db.get_script_batch(int_or_ext, batch_count)?;
+                let result: Vec<Vec<GetHistoryRes>> =
+                    self.client.batch_script_get_history(&batch)?;
+                let max = result
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, v)| !v.is_empty())
+                    .map(|(i, _)| i as u32)
+                    .max();
+                if let Some(max) = max {
+                    last_used[i as usize] = max + batch_count * BATCH_SIZE;
+                };
+
+                let flattened: Vec<GetHistoryRes> = result.into_iter().flatten().collect();
+                trace!("{}/batch({}) {:?}", i, batch_count, flattened.len());
+
+                if flattened.is_empty() {
+                    break;
+                }
+
+                for el in flattened {
+                    if el.height >= 0 {
+                        heights_set.insert(el.height as u32);
+                        txid_height.insert(el.tx_hash, el.height as u32);
+                    }
+                    history_txs_id.insert(el.tx_hash);
+                }
+
+                batch_count += 1;
+            }
+        }
+
+        self.db.insert_index(Index::External, last_used[Index::External as usize])?;
+        self.db.insert_index(Index::Internal, last_used[Index::Internal as usize])?;
+        trace!("last_used: {:?}", last_used,);
+
+        let new_txs = self.sync_txs(&history_txs_id)?;
+        self.sync_headers(&heights_set)?;
+        self.sync_height(&txid_height)?;
+
+        if new_txs {
+            self.db.flush()?;
+        }
+
+        trace!("elapsed {}", start.elapsed().as_millis());
+
+        Ok(new_txs)
+    }
+
+    fn sync_headers(&mut self, heights_set: &HashSet<u32>) -> Result<(), Error> {
+        let heights_in_db = self.db.get_only_heights()?;
+        let heights_to_download: Vec<u32> =
```

Comment:
Shouldn't this also download all the headers in between?

Reply:
this headers are used only to compute transaction timestamp, for (optional) SPV verification all the headers are downloaded